### ISSUE TITLE
`map_entry` warning for `HashMap` suggests code that does not compile

### ISSUE THREAD

**Issue by connortsui20** [rust-lang/rust-clippy#14224]
### Summary

The `map_entry` clippy warning can suggest code that does not compile. See below.

### Lint Name

map_entry

### Reproducer

If clippy sees an `contains_key` followed by an `insert` on `HashMap`, it will automatically (and correctly) suggest using `entry()` instead. However, if doesn't seem to take into account all branches. and can suggest code that does not compile.

```rust
pub fn insert_42_into_multimap(map: &mut HashMap<String, Vec<i32>>, key: String) {
    if !map.contains_key(&key) {
        map.get_mut(&key).unwrap().push(42);
    } else {
        map.insert(key, vec![42]);
    }
}
```

Note that this code is definitely not idiomatic, but the reason I know about this is that a student of mine (new to Rust) wrote something similar to the above code and got something similar to the following error message:

```
warning: usage of `contains_key` followed by `insert` on a `HashMap`
 --> src/lib.rs:4:5
  |
4 | /     if !map.contains_key(&key) {
5 | |         map.get_mut(&key).unwrap().push(42);
6 | |     } else {
7 | |         map.insert(key, vec![42]);
8 | |     }
  | |_____^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
  = note: `#[warn(clippy::map_entry)]` on by default
help: try
  |
4 ~     if let std::collections::hash_map::Entry::Occupied(mut e) = map.entry(key) {
5 +         e.insert(vec![42]);
6 +     } else {
7 +         map.get_mut(&key).unwrap().push(42);
8 +     }
  |
```

This suggestion is incorrect because `key` is moved and the other branch tries to borrow `&key`, thus this function will not compile.

I think the issue might be that if the `insert` following the `contains_key` is in a specific branch, clippy does not consider the other branches.

I'm wondering if it is possible to make clippy smart enough to suggest something like this (which would be the "idiomatic" way to write this code):

```rust
pub fn insert_42_into_multimap(map: &mut HashMap<String, Vec<i32>>, key: String) {
    map.entry(key).or_default().push(42);
}
```

I'm not super familiar with the internals of clippy but I'm assuming that this might be pretty hard. So maybe instead just make sure that it doesn't suggest code that is never going to compile in this case? Honestly not too sure what the solution should be here. 

However, I would still treat this seriously because the entry API is not the most immediately intuitive to new Rust programmers, and I think clippy should do its best to not suggest things that will give new Rustaceans more headaches than they probably already have (though of course I understand this is super hard).

### Version

```text
rustc 1.84.1 (e71f9a9a9 2025-01-27)
binary: rustc
commit-hash: e71f9a9a98b0faf423844bf0ba7438f29dc27d58
commit-date: 2025-01-27
host: x86_64-pc-windows-msvc
release: 1.84.1
LLVM version: 19.1.5

(This isn't platform specific, I ran the above code on the playground + on Linux)
```

**Comment by connortsui20** [comment]
Whoops this is a duplicate of #9305, #9470, #9925, #11976, #13306, and #13934

**Comment by jespiron** [comment]
Alas, like Napoleon, one does not simply invade `rust-lang` in the winter
but like Napoleon, better battles will be won
